torch/lib/THD/base/data_channels/Store.cpp

```diff
@@ -0,0 +1,178 @@
+#include "Store.hpp"
+#include "../ChannelUtils.hpp"
+
+#include <poll.h>
+#include <system_error>
+
+namespace thd {
+
+using store_type = std::unordered_map<std::string, std::vector<char>>;
+
+static void store_thread_daemon();
+static bool query(int socket);
+static bool check(const store_type& store,
+                  const std::vector<std::string>& keys);
+static void wake_up(int socket);
+
+enum class query_type : uint8_t {
+  set,
+  get,
+  wait,
+  finish,
+  stop_waiting
+};
+
+Store::Store()
+    : _rank(load_rank_env()) {
+  // Master runs the store_thread
+  if (_rank == 0) {
+    _store_thread = std::thread(store_thread_daemon);
+    std::tie(_master_port, std::ignore) = load_master_env();
+    _master_addr = "localhost";
+  } else {
+    std::tie(_master_addr, _master_port) = load_worker_env();
+  }
+  _socket = connect(_master_addr, _master_port);
+  send_bytes<rank_type>(_socket, &_rank, 1);
+}
+
+Store::~Store() {
+  // the 0 process has to stop the daemon
+  if (_rank == 0) {
+    send_value<query_type>(_socket, query_type::finish);
+    _store_thread.join();
+  }
+}
+
+void Store::set(const std::string& key, const std::vector<char>& data) {
+  send_value<query_type>(_socket, query_type::set);
+  send_string(_socket, key, true);
+  send_vector<char>(_socket, data);
+}
+
+std::vector<char> Store::get(const std::string& key) {
+  wait({key});
+  send_value<query_type>(_socket, query_type::get);
+  send_string(_socket, key);
+  return recv_vector<char>(_socket);
+}
+
+void Store::wait(const std::vector<std::string>& keys) {
+  send_value<query_type>(_socket, query_type::wait);
+  size_type nkeys = keys.size();
+  send_bytes<size_type>(_socket, &nkeys, 1, (nkeys > 0));
+  for (int i = 0; i < nkeys; i++) {
+    send_string(_socket, keys[i], (i != (nkeys - 1)));
+  }
+  // after sending the query, wait for a 'stop_waiting' reponse
+  query_type qr;
+  recv_bytes<query_type>(_socket, &qr, 1);
+  if (qr != query_type::stop_waiting)
+    throw std::runtime_error("stop_waiting response expected");
+}
+
+static void store_thread_daemon() {
+  port_type port;
+  rank_type world_size;
+  int socket;
+  std::tie(port, world_size) = load_master_env();
+  std::vector<int> sockets(world_size);
+  std::tie(socket, std::ignore) = listen(port);
+  // accept WORLD_SIZE connections
+  for (int i = 0; i < world_size; i++) {
+    int p_socket;
+    rank_type p_rank;
+    std::tie(p_socket, std::ignore) = accept(socket);
+    recv_bytes<rank_type>(p_socket, &p_rank, 1);
```

Comment:
It's not discarded, but truly the order doesn't matter, so `p_rank` needn't be send or received